google-ads-examples/src/main/java/com/google/ads/googleads/examples/misc/NavigateSearchResultPagesCachingTokens.java

```
@@ -0,0 +1,247 @@
+    // The first page's token is always an empty string.
+    List<String> pageTokens = new ArrayList<>(Collections.singletonList(""));
+
+    System.out.println("--- 0. Fetch page 1 to get metadata:");
+
+    // Creates a query that retrieves the campaigns.
+    String query =
+        String.format(
+            "SELECT campaign.id, campaign.name FROM campaign ORDER BY campaign.name LIMIT %d",
+            RESULTS_LIMIT);
+
+    // Creates a paginated search request.
+    SearchGoogleAdsRequest request =
+        SearchGoogleAdsRequest.newBuilder()
+            .setCustomerId(Long.toString(customerId))
+            .setPageSize(PAGE_SIZE)
+            .setQuery(query)
+            .setReturnTotalResultsCount(true)
+            .build();
+
+    int totalNumberOfPages;
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      // Issues a paginated search request.
+      SearchPagedResponse response = googleAdsServiceClient.search(request);
+      cacheNextPageToken(pageTokens, response.getPage(), 1);
+
+      // Determines the total number of results and prints it.
+      // The total results count does not take into consideration the LIMIT clause of the query,
+      // so we need to find the minimal value between the limit and the total results count.
+      long totalNumberOfResults =
+          Math.min(RESULTS_LIMIT, response.getPage().getResponse().getTotalResultsCount());
+      System.out.printf("Total number of campaigns found: %d.%n", totalNumberOfResults);
+
+      // Determines the total number of pages and prints it.
+      totalNumberOfPages = (int) Math.ceil((double) totalNumberOfResults / (double) PAGE_SIZE);
+      System.out.printf("Total number of pages: %d.%n", totalNumberOfPages);
+      if (totalNumberOfPages == 0) {
+        System.out.println("Could not find any campaigns.");
+        System.exit(1);
```

Prefer `return` over `System.exit`. We should only do the latter in `main` so if someone copies and pastes this code they won't inadvertently shut down their JVM on an empty response.

If detecting an empty response is important (e.g., for FT), then you could just throw an exception here such as `IllegalStateException`.